src/Transform/ONNX/Decompose.cpp

```diff
@@ -199,6 +199,52 @@ void populateDecomposingONNXBeforeMhloPatterns(
 }
 #endif
 
+// Special Op fusion for the following pattern:
+//   %1 = Concat(inputs, axis)
+//   %2 = Shape(%1, start, end)
+//   %3 = Transpose(%1, perm)
+// into a special Op
+//   %2, %3 = ConcatShapeTranspose(inputs, axis, start, end, perm)
+// This fusion is an experimental work for performance
+struct ConcatFusePattern : public ConversionPattern {
+  ConcatFusePattern(MLIRContext *context)
+      : ConversionPattern(ONNXConcatOp::getOperationName(), 4, context) {}
+  LogicalResult matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const final {
+
+    ONNXConcatOp concatOp = ::llvm::dyn_cast<ONNXConcatOp>(op);
+
+    // Match
+    ONNXShapeOp shapeOp = NULL;
+    ONNXTransposeOp transposeOp = NULL;
+    bool failed = false;
+    // Check there are exactly two users by Shape and Transpose
+    for (Operation *user : op->getUsers()) {
+      if (isa<ONNXShapeOp>(user) && !shapeOp)
+        shapeOp = cast<ONNXShapeOp>(user);
+      else if (isa<ONNXTransposeOp>(user) && !transposeOp)
+        transposeOp = cast<ONNXTransposeOp>(user);
+      else
+        failed = true;
+    }
+    if (!(shapeOp && transposeOp && !failed))
+      return success();
```

Comment:
Just curious, should this be `return failure()` to say the pattern does not match.

Reply:
Yes, it should be failure.

Reply:
Actually I am not sure whether the failure means failure of match or rewrite. Maybe for match. If match and write is defined separately, the rewrite does not have return value.  If there are more than one rule for one operation, the match failure for each rule should not cause the failure the applyxxxxConversion.